server/src/main/java/org/elasticsearch/action/admin/indices/forcemerge/TransportForceMergeAction.java

```diff
@@ -92,12 +93,16 @@ protected void shardOperation(
         ActionListener<TransportBroadcastByNodeAction.EmptyResult> listener
     ) {
         assert (task instanceof CancellableTask) == false; // TODO: add cancellation handling here once the task supports it
-        threadPool.executor(ThreadPool.Names.FORCE_MERGE).execute(ActionRunnable.supply(listener, () -> {
```

Comment:
nit May I ask where does it filter out the unpromotable shards? I was trying to figure it out but gave up after a couple of minutes.

Reply:
It doesn't https://github.com/elastic/elasticsearch/blob/44ecfc2d8c44b98b84d746582190fc8b150aaca9/server/src/main/java/org/elasticsearch/action/admin/indices/forcemerge/TransportForceMergeAction.java#L108-L114, it's just a no-op in the `SearchEngine`.